lib/get-repo-pipeline-manager.js

```diff
@@ -0,0 +1,204 @@
+  pushPipeline.addMiddleware('failed-to-push-error', async (next, repository, branchName, options) => {
+    try {
+      const result = await next();
+      return result;
+    } catch (error) {
+      if (!(error instanceof GitError)) { throw error; }
+      if (/rejected[\s\S]*failed to push/.test(error.stdErr)) {
+        notificationManager.addError('Push rejected', {
+          description: 'The tip of your current branch is behind its remote counterpart.' +
+            ' Try pulling before pushing again. Or, to force push, hold `cmd` or `ctrl` while clicking.',
+          dismissable: true,
+        });
+      } else {
+        notificationManager.addError('Unable to push', {
+          description: `<pre>${error.stdErr}</pre>`,
+          dismissable: true,
+        });
+      }
+      throw error;
+    }
+  });
+
+  const pullPipeline = pipelineManager.getPipeline(pipelineManager.actionKeys.PULL);
+  pullPipeline.addMiddleware('set-pull-in-progress', async (next, repository, branchName) => {
+    repository.getOperationStates().setPullInProgress(true);
+    try {
+      await next();
+    } finally {
+      repository.getOperationStates().setPullInProgress(false);
+    }
+  });
+  pullPipeline.addMiddleware('failed-to-pull-error', async (next, repository, branchName) => {
+    try {
+      const result = await next();
+      return result;
+    } catch (error) {
+      if (!(error instanceof GitError)) { throw error; }
+      if (/error: Your local changes to the following files would be overwritten by merge/.test(error.stdErr)) {
+        const lines = error.stdErr.split('\n');
+        const files = lines.slice(3, lines.length - 3).map(l => `\`${l.trim()}\``).join('<br>');
+        notificationManager.addError('Pull aborted', {
+          description: 'Local changes to the following would be overwritten by merge:<br>' + files +
+            '<br>Please commit your changes or stash them before you merge.',
+          dismissable: true,
+        });
+      } else if (/Automatic merge failed; fix conflicts and then commit the result./.test(error.stdOut)) {
+        repository.didMergeError();
+        notificationManager.addWarning('Merge conflicts', {
+          description: `Your local changes conflicted with changes made on the remote branch. Resolve the conflicts
+            with the Git panel and commit to continue.`,
+          dismissable: true,
+        });
+      } else {
+        notificationManager.addError('Unable to pull', {
+          description: `<pre>${error.stdErr}</pre>`,
+          dismissable: true,
+        });
+      }
+      throw error;
+    }
+  });
+
+  const fetchPipeline = pipelineManager.getPipeline(pipelineManager.actionKeys.FETCH);
+  fetchPipeline.addMiddleware('set-fetch-in-progress', async (next, repository) => {
+    repository.getOperationStates().setFetchInProgress(true);
+    try {
+      await next();
+    } finally {
+      repository.getOperationStates().setFetchInProgress(false);
+    }
+  });
```

Comment:
Yeah, it would make things much less repetitive and I do like "clever". My small concern is that would make it harder to grep through the code to find all places where `setThingInProgress` is used. And if this file gets separated into multiple files for each operation (say after it grows a bunch), then we'd have to pull in `createInProgressMiddleware` from a different file and it's less immediately clear from glancing at the code what is happening.